app/js/middlewares/responseHandlerMiddleware.js

```diff
@@ -0,0 +1,69 @@
+import {
+  GET_DATE_SUCCESS,
+  GET_DATE_FAILURE,
+  FETCH_ENCOUNTER_ROLE_SUCCESS,
+  FETCH_ENCOUNTER_ROLE_FAILURE,
+  FETCH_ENCOUNTER_TYPE_SUCCESS,
+  FETCH_ENCOUNTER_TYPE_FAILURE,
+  LOAD_PAST_ORDERS_SUCCESS,
+} from '../actions/actionTypes';
+
+export default function responseHandlerMiddleware() {
+  return next => (action) => {
+    if (action.type === GET_DATE_SUCCESS) {
+      const { results } = action.payload.data;
+      try {
+        if (!results.length) {
+          const DATE_FORMAT = 'DD-MMM-YYYY HH:mm';
+          return next({ ...action, dateFormat: DATE_FORMAT });
+        } else if (results.length && results[0].value === null) {
+          throw Error('incomplete config');
+        } else {
+          return next({ ...action, dateFormat: results[0].value });
+        }
+      } catch (error) {
+        console.warn(`${GET_DATE_FAILURE}-${error}`); /*eslint-disable-line*/
+        return next({ type: GET_DATE_FAILURE, payload: error });
+      }
+    }
+
+    if (
+      action.type === FETCH_ENCOUNTER_ROLE_SUCCESS
+      || action.type === FETCH_ENCOUNTER_TYPE_SUCCESS
+    ) {
+      const { results } = action.payload.data;
+      try {
+        if (results.length === 0) {
+          throw Error('incomplete config');
+        } else {
+          return next({ ...action, data: results[0] });
+        }
+      } catch (error) {
+        console.warn(`${FETCH_ENCOUNTER_ROLE_FAILURE || FETCH_ENCOUNTER_TYPE_FAILURE}-${error}`); /*eslint-disable-line*/
+        return next({ type: FETCH_ENCOUNTER_ROLE_FAILURE, payload: error });
+      }
+    }
+
+    if (action.type === LOAD_PAST_ORDERS_SUCCESS) {
+      const { results, totalCount } = action.payload.data;
+      const { limit, startIndex } = action.meta;
+      const pageCount = Math.ceil(totalCount / limit);
+      const startIndexLimit = startIndex + limit;
+      const from = startIndex + 1;
+      const to = startIndexLimit > totalCount ? totalCount : startIndexLimit;
+      const pastOrdersResultCount = `Showing ${from} to ${to} of ${totalCount} entries`;
+      next({ type: 'PAST_ORDERS_RESULT_COUNT', pastOrdersResultCount });
+      next({ type: 'PAST_ORDERS_PAGE_COUNT', pageCount });
+      return next({ ...action, pastOrders: results });
+    }
+    if (action.payload) {
+      return next({
+        ...action,
+        type: action.type,
+        data: action.payload.data,
+        meta: action.meta || {},
+      });
+    }
+    return next(action);
+  };
+}
```

Comment:
@rhenshaw56  I'm still learning how all this works... can you explain what the theory is behind this middleware?  That is, how do we determine what logic goes here as opposed to in a reducer?  Am I right that this seems to handle the kind of stuff that would go in a saga if we were using sagas in this project?

Reply:
@mogoodrich not really! so in sagas you're watching for events to dispatch actions which is different in this case. so what happens is;
```
const middlewares = [
  thunk,
  sagaMiddleware,
  errorMiddleware,
  promiseMiddleware({ promiseTypeSuffixes }),
  responseHandlerMiddleware,
];
```
firstly, the `thunk` middleware is declared first so it tries to handle everything that is a `thunk` ie any action that is coming as a function, resolves it and passes it on to the next middleware, but the saga middleware can't handle that because thunk will return an object instead the `saga` middleware watches for sagas and executes them and passes it on the next middleware which is the errorMiddleware`, here we're just trying to catch errors resulting from failed promises and then passing them on to the next middleware which is the `promiseMiddleware` that handles pending promises and resolves unrejected ones. the `responseHandler` middleware is just for formatting responses from the server and the reason why it's placed after the `promiseMiddleware` is that it only handles those being passed down by the `promiseMiddleware` after thy've been resolved cos they come in as promises. so I noticed that in some actions we were returning `response.data`, `response.data.results`, and in most cases, we'd have to format it in some way before it gets to the reducer.

Reply:
Got is, so the "responseHandlerMiddleware" is generally responsible for making sure that the async responses get formatting into some sort of standard format?  When I'm adding a new feature that fetching and manipulating data, just trying to figure out how I determine what logic should go in the responseHandlerMiddleware and what should go in a reducer.

@rhenshaw56

Reply:
@mogoodrich yeah, that's right and if you check line 59 of the responsHandlerMiddleware, every request made henceforth will by default be gotten as action.data in the every reducer

Reply:
Sounds good!